applications/CompressiblePotentialFlowApplication/custom_processes/define_2d_wake_process.cpp

```diff
@@ -0,0 +1,299 @@
+    double projection_on_wake = inner_prod(distance_to_element_center, mWakeDirection);
+
+    return projection_on_wake > 0.0;
+}
+
+// This function computes the distance of the element nodes to the wake
+const BoundedVector<double, 3> Define2DWakeProcess::ComputeNodalDistancesToWake(ElementIteratorType& rElement)
+{
+    BoundedVector<double, 3> nodal_distances_to_wake = ZeroVector(3);
+    BoundedVector<double, 3> distance_from_te_to_node = ZeroVector(3);
+    double distance_to_wake = 1.0;
+    for (unsigned int i = 0; i < rElement->GetGeometry().size(); i++) {
+        // Compute the distance from the trailing edge to the node
+        distance_from_te_to_node =
+            ComputeDistanceFromTrailingEdgeToPoint(rElement->GetGeometry()[i]);
+
+        // Compute the projection of the distance vector in the wake normal
+        // direction
+        distance_to_wake = inner_prod(distance_from_te_to_node, mWakeNormal);
+
+        // Nodes laying on the wake have a positive distance
+        if (std::abs(distance_to_wake) < mTolerance) {
+            distance_to_wake = mTolerance;
+        }
+        nodal_distances_to_wake[i] = distance_to_wake;
+    }
+    return nodal_distances_to_wake;
+}
+
+// This function checks whether the element is cut by the wake
+const bool Define2DWakeProcess::CheckIfWakeElement(BoundedVector<double, 3>& rNodalDistancesToWake)
+{
+    // Initialize counters
+    unsigned int number_of_nodes_with_positive_distance = 0;
+    unsigned int number_of_nodes_with_negative_distance = 0;
+
+    // Count how many element nodes are above and below the wake
+    for (unsigned int i = 0; i < rNodalDistancesToWake.size(); i++) {
+        if (rNodalDistancesToWake(i) < 0.0) {
+            number_of_nodes_with_negative_distance += 1;
+        }
+        else {
+            number_of_nodes_with_positive_distance += 1;
+        }
+    }
+
+    // Elements with nodes above and below the wake are wake elements
+    return number_of_nodes_with_negative_distance > 0 &&
+           number_of_nodes_with_positive_distance > 0;
+}
+
+// This function adds the trailing edge elements in the
+// trailing_edge_sub_model_part
+const void Define2DWakeProcess::AddTrailingEdgeElements()
+{
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+    std::sort(mTrailingEdgeElementsOrderedIds.begin(),
+              mTrailingEdgeElementsOrderedIds.end());
+    root_model_part.GetSubModelPart("trailing_edge_sub_model_part").AddElements(mTrailingEdgeElementsOrderedIds);
+}
+
+// This function selects the kutta elements. Kutta elements are touching the
+// trailing edge from below
+const void Define2DWakeProcess::MarkKuttaElements()
+{
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+    ModelPart& trailing_edge_sub_model_part =
+        root_model_part.GetSubModelPart("trailing_edge_sub_model_part");
+
+    BoundedVector<double, 3> distance_to_element_center = ZeroVector(3);
+    double distance_to_wake = 1.0;
+
```

Comment:
I think this loop can be parallelized.

Reply:
You are right, in can be parallelized. The trailing_edge_sub_model_part has however a maximum of 8 elements so it doesn't really make a difference in terms of performance...

Reply:
then don't do it otherwise you slow it down